src/request.rs

```diff
@@ -238,27 +267,176 @@ impl RequestBuilder {
     }
 }
 
+impl MultipartRequestBuilder {
+    /// Constructs a new request.
+    pub fn new(request_builder: RequestBuilder) -> MultipartRequestBuilder {
+        MultipartRequestBuilder { request_builder: request_builder, params: None, files: None }
+    }
+
+    /// Add a `Header` to this Request.
+    ///
+    /// ```rust
+    /// # use reqwest::Error;
+    /// #
+    /// # fn run() -> Result<(), Error> {
+    /// use reqwest::header::UserAgent;
+    /// let client = reqwest::Client::new()?;
+    ///
+    /// let res = client.get("https://www.rust-lang.org")?
+    ///     .header(UserAgent("foo".to_string()))
+    ///     .send()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn header<H>(&mut self, header: H) -> &mut MultipartRequestBuilder
+        where H: ::header::Header + ::header::HeaderFormat
+    {
+        self.request_builder.request_mut().headers.set(header);
+        self
+    }
+
+    /// Add a set of Headers to the existing ones on this Request.
+    ///
+    /// The headers will be merged in to any already set.
+    pub fn headers(&mut self, headers: ::header::Headers) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().headers.extend(headers.iter());
+        self
+    }
+
+    /// Enable HTTP basic authentication.
+    pub fn basic_auth<U, P>(&mut self, username: U, password: Option<P>) -> &mut MultipartRequestBuilder
+        where U: Into<String>,
+              P: Into<String>
+    {
+        self.request_builder.header(::header::Authorization(::header::Basic {
+                                                username: username.into(),
+                                                password: password.map(|p| p.into()),
+                                            }));
+        self
+    }
+
+    /// Set the request body.
+    pub fn body<T: Into<Body>>(&mut self, body: T) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().body = Some(body.into());
+        self
+    }
+
+    /// Add parameters to the multipart/form-data `Request`.
+    pub fn params(&mut self, params: Params) -> &mut MultipartRequestBuilder {
```

Comment:
@seanmonstar When making a multipart/form-data request you'd ideally want to have the parameters on top of the request and the files beneath it. In your `RequestBuilder.form` method you write the body directly because there isn't anything else to write. However, in the `MultipartRequestBuilder` you cannot write these directly because it could mess up the order of the things. Which means I need to store the params and files and write them all in the right order when the `build` method is called. With `Serialize` I could store the given params as JSON. But it seems like a lot of overhead. The `params` method as it now is seems to me the most straightforward way with the least overhead, I could maybe add  a `param` method and a `params` method just like with the files. Let me know what your thoughts are on this.